File: src/lib/agents/codex.ts
import { BaseAgent, type StopSignal } from './base.js';
import type { AgentConfig, ApprovalMode } from '../config/schema.js';

const APPROVAL_FLAGS: Record<ApprovalMode, string[]> = {
  'full-auto': ['--full-auto'],
  'auto-edit': ['-a', 'on-request', '-s', 'workspace-write'],
  'suggest': ['-a', 'untrusted', '-s', 'read-only'],
  'default': [],
};

export class CodexAgent extends BaseAgent {
  constructor(config: AgentConfig, workDir = '.', approvalMode: ApprovalMode = 'default') {
    super(config, workDir, approvalMode);
  }

  buildCmd(): string[] {
    const cmd = ['codex', 'exec', '--skip-git-repo-check'];
    cmd.push(...(APPROVAL_FLAGS[this.approvalMode] ?? []));
    if (this.config.model) cmd.push('--model', this.config.model);
    return cmd;
  }

  async healthCheck(): Promise<boolean> {
    try { await this.runCli(['codex', '--version']); return true; }
    catch { return false; }
  }

  async send(message: string): Promise<string> {
    return this.execute(message);
  }

  async execute(task: string, context = ''): Promise<string> {
    const prompt = context ? `${task}\n\n${context}` : task;
    const raw = await this.runCli([...this.buildCmd(), prompt]);
    return this.extractResult(raw);
  }

  override async runCliStreaming(
    args: string[],
    onLine?: (line: string) => StopSignal | null | undefined,
    noOutputTimeout?: number,
  ): Promise<string> {
    const raw = await super.runCliStreaming(args, onLine, noOutputTimeout);
    return this.extractResult(raw);
  }

  // Codex outputs a structured log ending with:
  //   tokens used\n<count>\n<clean final response>
  // Everything before that marker (headers, exec traces, warnings) is stripped.
  private extractResult(raw: string): string {
    const marker = '\ntokens used\n';
    const idx = raw.lastIndexOf(marker);
    if (idx !== -1) {
      const after = raw.slice(idx + marker.length);
      // First line is the numeric count ("2,282"), the rest is the clean response
      const response = after.split('\n').slice(1).join('\n').trim();
      if (response) return response;
    }
    return raw;
  }
}

File: src/swarm/agents/codex.py
"""Adapter for OpenAI Codex CLI (codex)."""

from __future__ import annotations

from swarm.config.schema import AgentConfig, AgentType, ApprovalMode

from .base import BaseAgent

_APPROVAL_FLAGS: dict[ApprovalMode, list[str]] = {
    ApprovalMode.FULL_AUTO: ["--full-auto"],
    ApprovalMode.AUTO_EDIT: ["--auto-edit"],
    ApprovalMode.SUGGEST: ["--suggest"],
    ApprovalMode.DEFAULT: [],
}


class CodexAgent(BaseAgent):
    """Wraps the `codex` CLI."""

    agent_type = AgentType.CODEX

    def _build_cmd(self) -> list[str]:
        cmd = ["codex", "--quiet"]
        cmd.extend(_APPROVAL_FLAGS.get(self.approval_mode, []))
        if self.config.model:
            cmd.extend(["--model", self.config.model])
        return cmd

    async def health_check(self) -> bool:
        try:
            await self._run_cli(["codex", "--version"])
            return True
        except Exception:
            return False

    async def execute(self, title: str, description: str) -> str:
        prompt = f"{title}\n\n{description}" if description else title
        return await self._run_cli([*self._build_cmd(), prompt])

    async def decompose(self, prompt: str) -> list[dict]:
        return [{"title": prompt, "description": ""}]

    async def synthesize(self, original_prompt: str, results: dict[str, str]) -> str:
        results_text = "\n".join(f"- {tid}: {r[:200]}" for tid, r in results.items())
        prompt = f"Summarize these results for: {original_prompt}\n\n{results_text}"
        return await self._run_cli([*self._build_cmd(), prompt])
